core/src/main/java/org/jruby/RubyArray.java

```diff
@@ -2047,12 +2047,17 @@ public IRubyObject to_h(ThreadContext context) {
         RubyHash hash = useSmallHash ? RubyHash.newSmallHash(runtime) : RubyHash.newHash(runtime);
 
         for (int i = 0; i < realLength; i++) {
-            IRubyObject elt = eltInternal(i).checkArrayType();
-            if (elt == context.nil) {
-                throw context.runtime.newTypeError("wrong element type " + eltInternal(i).getMetaClass().getRealClass() + " at " + i + " (expected array)");
+            IRubyObject e = eltInternal(i);
+            IRubyObject elt, key_value_pair;
```

Comment:
Could you move decls to their first assignment:

```java
  IRubyObject elt = block.isGiven() // ...
```

Reply:
Fixed.